### DISCUSSION TITLE
Issue with TypeORM decorators when running vitest in TypeScript project

### DISCUSSION THREAD

**Discussion by carlosmoliv** [vitest-dev/vitest#3320] [Q&A] [answered]
I'm working with TypeORM and I noticed that my code is not working when I don't explicitly set the column type, for example @Column({ type: 'varchar' }). 

I've already added "experimentalDecorators": true and "emitDecoratorMetadata": true to the tsconfig.json file and imported reflect-metadata at the entry point of the API, but the issue persists. Can anyone suggest what else I could be missing or any potential solutions to this problem?

![image](https://user-images.githubusercontent.com/52833750/236880508-5135a4be-4288-42b2-9807-b1ff438d8890.png)

The typeorm entity:

```
@Entity({ name: 'addresses' })
export class AddressDbModel {
  @Column()
  street: string;

  @Column()
  stretNumber: number;

  @Column()
  neighborhood: string;

  @Column()
  city: string;

  @Column()
  state: string;

  @Column()
  zipCode: string;

  @Column({ nullable: true })
  complement: string;

  @Column({ nullable: true })
  country: string;
}

```

**Comment by sheremet-va** [comment]
`emitDecoratorMetadata` is not supported by esbuild: https://github.com/evanw/esbuild/issues/257

**Comment by thiagomini** [comment, marked as answer]
Hey, @carlosmoliv. Install the following package: `unplugin-swc`

then, add this configuration to the vitest config file:

```ts
import swc from 'unplugin-swc';

export default defineConfig({
  test: {
   // ... configs
  },
  // add the line below
  plugins: [swc.vite()],
});

**Reply by carlosmoliv** [reply]
Hi @thiagomini , last week I tried using that plugin to fix another issue I had, but it didn't work. Here's the link to the discussion: https://github.com/vitest-dev/vitest/discussions/3290

But I'm gonna try it now and see if it can at least fix the problem with decorators

**Reply by carlosmoliv** [reply]
@thiagomini It solved the issue with the decorators! Thanks

**Reply by eremzeit** [reply]
I tried using the `unplugin-swc` approach and it fixed the exception thrown talking about the `emitDecoratorMetadata` flag, but then when I ran my code in Vitest I'm noticing that it's as if the decorators aren't being run.  As in, the entity values that are supposed to be autogenerated aren't being autogenerated.  I haven't found the fix yet.

**Reply by thiagomini** [reply]
I suggest you to create a minimum reproducible repo that isolates this issue, eremzeit. There could be many reasons for your problem! For instance, you have to ensure you are using the correct options for `swc.vite()` - if you have a `.swcrc` file it could be overriding the values there. Not sure if you are using Nest, but this might serve in your case as well: https://docs.nestjs.com/recipes/swc#vitest

**Reply by eremzeit** [reply]
I figured out the issue, but when I came back to write about it, I realized this thread is talking about TypeORM and my issue is with TypeDORM (facepalm).  Sorry for the trouble!

**Reply by ValentinGurkov** [reply]
@eremzeit Can you share what was the reason the decorators were not executing? I'm facing the same thing...

**Reply by byrpatrick** [reply]
Hey @thiagomini thanks for this suggestion, it fixed an error I was having when writing a test for a Next API Route using [Next API Decorators](https://github.com/instantcommerce/next-api-decorators) and [Node Mocks Http](https://github.com/eugef/node-mocks-http). 💯

**Reply by KiddoV** [reply]
Not working with my ``Svelte + Vite`` project. When added ``plugins: [swc.vite()],``, the page just came blank and no errors popup.

**Reply by okanaslan** [reply]
I get the following error when I try to run Typeorm tests with Vitest & unplugin-swc package.

```bash
[Nest] 69877  - 06/24/2025, 10:23:25 PM   ERROR [TypeOrmModule] Unable to connect to the database. Retrying (3)...
SyntaxError: Unexpected identifier 'Relation'
    at compileSourceTextModule (node:internal/modules/esm/utils:344:16)
    at ModuleLoader.importSyncForRequire (node:internal/modules/esm/loader:420:18)
    at loadESMFromCJS (node:internal/modules/cjs/loader:1561:24)
    at Module._compile (node:internal/modules/cjs/loader:1712:5)
    at Object..js (node:internal/modules/cjs/loader:1895:10)
    at Module.load (node:internal/modules/cjs/loader:1465:32)
    at Function._load (node:internal/modules/cjs/loader:1282:12)
    at TracingChannel.traceSync (node:diagnostics_channel:322:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:235:24)
    at Module.require (node:internal/modules/cjs/loader:1487:12)
```

**Reply by linonetwo** [reply]
Thank you! This works for me!
